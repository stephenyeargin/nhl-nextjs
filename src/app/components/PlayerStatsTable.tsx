import React from 'react';
import TeamLogo from '@/app/components/TeamLogo';
import statsStyles from '@/app/components/StatsTable.module.scss';
import { formatSeason, formatStat } from '@/app/utils/formatters';

type StatHeader = {
  key: string;
  label: string;
  title: string;
  altKey?: string;
  precision?: number;
};

export interface PlayerSeasonTotals {
  season: string | number;
  leagueAbbrev: string;
  gameTypeId: number;
  teamName?: { default?: string };
  [stat: string]: string | number | { default?: string } | undefined;
}

interface PlayerStatsTableProps {
  stats: PlayerSeasonTotals[];
  statHeaders: readonly StatHeader[];
  showLeague: boolean;
  headerColorClass: string;
  headerStyle?: React.CSSProperties;
}

const PlayerStatsTable: React.FC<PlayerStatsTableProps> = ({
  stats,
  statHeaders,
  showLeague,
  headerColorClass,
  headerStyle,
}) => {
  const visibleHeaders = statHeaders.filter((h) => {
    const key = h.key as string;
    const altKey = h.altKey;
    const present = stats.some(
      (season) =>
        Object.prototype.hasOwnProperty.call(season, key) ||
        (altKey && Object.prototype.hasOwnProperty.call(season, altKey))
    );

    return present;
  });

  return (
    <div className="overflow-x-auto">
      <table className={statsStyles.statsTable}>
        <thead>
          <tr className={`text-xs border ${headerColorClass}`}>
            <th className={'p-2 text-center'} style={headerStyle}>
              Season
            </th>
            <th className={'p-2 text-left'} style={headerStyle}>
              Team
            </th>
            {showLeague && (
              <th className={'p-2 text-left'} style={headerStyle}>
                League
              </th>
            )}
            {visibleHeaders.map((h) => {
              const key = h.key;
              const label = h.label;
              const title = h.title as string;

              return (
                <th key={key} className={`p-2 text-center ${headerColorClass}`} style={headerStyle}>
                  <abbr className="underline decoration-dashed" title={title}>
                    {label}
                  </abbr>
                </th>
              );
            })}
          </tr>
        </thead>
        <tbody>
          {stats.map((season, i) => (
            <tr key={i} className={`${i % 2 ? 'bg-slate-500/10' : ''}`}>
              <td className="p-2 border text-center">{formatSeason(season.season)}</td>
              <td className="">
                <div className="flex gap-1 items-center">
                  {season.leagueAbbrev === 'NHL' && (
                    <TeamLogo
                      team={(season.teamName as any)?.default}
                      className="h-8 w-8 hidden md:block"
                      alt={String((season.teamName as any)?.default || '')}
                    />
                  )}
                  {(season.teamName as any)?.default}
                </div>
              </td>
              {showLeague && <td className="text-left">{season.leagueAbbrev}</td>}
              {visibleHeaders.map((h) => {
                const key = h.key;
                const altKey = h.altKey;
                const precision = h.precision;
                const primaryVal = season[key];
                const altVal = altKey ? season[altKey] : undefined;
                const value =
                  typeof primaryVal === 'number' || typeof primaryVal === 'string'
                    ? primaryVal
                    : typeof altVal === 'number' || typeof altVal === 'string'
                      ? altVal
                      : undefined;

                const content = (() => {
                  if (value === undefined) {
                    return '--';
                  }
                  if (key === 'plusMinus') {
                    const n = typeof value === 'number' ? value : Number(value);
                    if (Number.isNaN(n)) {
                      return '--';
                    }

                    return n > 0 ? `+${n}` : `${n}`;
                  }

                  return formatStat(value as any, precision);
                })();

                return (
                  <td key={key} className="p-2 border text-center text-xs">
                    {content}
                  </td>
                );
              })}
            </tr>
          ))}
        </tbody>
      </table>
    </div>
  );
};

export default PlayerStatsTable;
