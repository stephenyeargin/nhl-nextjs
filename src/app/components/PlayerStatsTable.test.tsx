import React from 'react';
import { render, screen, within } from '@testing-library/react';
import type { PlayerSeasonTotals } from '@/app/components/PlayerStatsTable';
import PlayerStatsTable from '@/app/components/PlayerStatsTable';

const headers = [
  { key: 'gamesPlayed', label: 'GP', title: 'Games Played' },
  { key: 'goals', label: 'G', title: 'Goals' },
  { key: 'plusMinus', label: '+/-', title: 'Plus/Minus' },
] as const;

const mkRow = (overrides: Partial<PlayerSeasonTotals>): PlayerSeasonTotals => ({
  season: '20192020',
  leagueAbbrev: 'NHL',
  gameTypeId: 2,
  teamName: { default: 'NSH' },
  ...overrides,
});

describe('PlayerStatsTable', () => {
  it('shows only columns present across dataset and keeps alignment with --', () => {
    const stats: PlayerSeasonTotals[] = [
      mkRow({ gamesPlayed: 82, goals: 10, plusMinus: 5 }),
      mkRow({ gamesPlayed: 70 }), // missing goals and plusMinus
    ];

    render(
      <PlayerStatsTable
        stats={stats}
        statHeaders={headers}
        showLeague={false}
        headerColorClass=""
      />
    );

    // Header cells
    expect(screen.getByText('Season')).toBeInTheDocument();
    expect(screen.getByText('Team')).toBeInTheDocument();
    // All three headers are visible because at least one row has each
    expect(screen.getByText('GP')).toBeInTheDocument();
    expect(screen.getByText('G')).toBeInTheDocument();
    expect(screen.getByText('+/-')).toBeInTheDocument();

    const rows = screen.getAllByRole('row');
    // First data row should contain 82, 10, +5
    const r1 = rows[1];
    expect(within(r1).getByText('82')).toBeInTheDocument();
    expect(within(r1).getByText('10')).toBeInTheDocument();
    expect(within(r1).getByText('+5')).toBeInTheDocument();

    // Second data row should render placeholders for missing stats
    const r2 = rows[2];
    expect(within(r2).getByText('70')).toBeInTheDocument();
    // goals placeholder
    expect(within(r2).getAllByText('--').length).toBeGreaterThan(0);
  });

  it('matches snapshot (basic NHL table)', () => {
    const stats: PlayerSeasonTotals[] = [
      mkRow({ gamesPlayed: 82, goals: 10, plusMinus: 5 }),
      mkRow({ gamesPlayed: 70 }),
    ];

    const { container } = render(
      <PlayerStatsTable
        stats={stats}
        statHeaders={headers}
        showLeague={false}
        headerColorClass=""
      />
    );

    expect(container.firstChild).toMatchSnapshot();
  });
});
