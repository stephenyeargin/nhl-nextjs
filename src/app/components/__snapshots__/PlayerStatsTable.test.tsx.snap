// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlayerStatsTable matches snapshot (basic NHL table) 1`] = `
<div
  class="overflow-x-auto"
>
  <table
    class="statsTable"
  >
    <thead>
      <tr
        class="text-xs border "
      >
        <th
          class="p-2 text-center"
        >
          Season
        </th>
        <th
          class="p-2 text-left"
        >
          Team
        </th>
        <th
          class="p-2 text-center "
        >
          <abbr
            class="underline decoration-dashed"
            title="Games Played"
          >
            GP
          </abbr>
        </th>
        <th
          class="p-2 text-center "
        >
          <abbr
            class="underline decoration-dashed"
            title="Goals"
          >
            G
          </abbr>
        </th>
        <th
          class="p-2 text-center "
        >
          <abbr
            class="underline decoration-dashed"
            title="Plus/Minus"
          >
            +/-
          </abbr>
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        class=""
      >
        <td
          class="p-2 border text-center"
        >
          2019-20
        </td>
        <td
          class=""
        >
          <div
            class="flex gap-1 items-center"
          >
            <a
              href="/team/NSH"
            >
              <img
                alt="NSH"
                class="h-8 w-8 hidden md:block"
                data-nimg="1"
                decoding="async"
                height="256"
                loading="lazy"
                src="https://assets.nhle.com/logos/nhl/svg/NSH_light.svg"
                style="color: transparent;"
                width="256"
              />
            </a>
            NSH
          </div>
        </td>
        <td
          class="p-2 border text-center text-xs"
        >
          82
        </td>
        <td
          class="p-2 border text-center text-xs"
        >
          10
        </td>
        <td
          class="p-2 border text-center text-xs"
        >
          +5
        </td>
      </tr>
      <tr
        class="bg-slate-500/10"
      >
        <td
          class="p-2 border text-center"
        >
          2019-20
        </td>
        <td
          class=""
        >
          <div
            class="flex gap-1 items-center"
          >
            <a
              href="/team/NSH"
            >
              <img
                alt="NSH"
                class="h-8 w-8 hidden md:block"
                data-nimg="1"
                decoding="async"
                height="256"
                loading="lazy"
                src="https://assets.nhle.com/logos/nhl/svg/NSH_light.svg"
                style="color: transparent;"
                width="256"
              />
            </a>
            NSH
          </div>
        </td>
        <td
          class="p-2 border text-center text-xs"
        >
          70
        </td>
        <td
          class="p-2 border text-center text-xs"
        >
          --
        </td>
        <td
          class="p-2 border text-center text-xs"
        >
          --
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
